libsolutil/CommonData.h

```diff
@@ -321,6 +336,44 @@ void joinMap(std::map<K, V>& _a, std::map<K, V>&& _b, F _conflictSolver)
 	}
 }
 
+namespace detail
+{
+
+template<typename Container, typename Value>
+auto findOffset(Container&& _container, Value&& _value, int)
+-> decltype(_container.find(_value) == _container.end(), std::distance(_container.begin(), _container.find(_value)), std::optional<size_t>())
+{
+	auto it = _container.find(std::forward<Value>(_value));
+	auto end = _container.end();
+	if (it == end)
+		return std::nullopt;
+	return std::distance(_container.begin(), it);
+}
+template<typename Range, typename Value>
+auto findOffset(Range&& _range, Value&& _value, void*)
+-> decltype(std::find(std::begin(_range), std::end(_range), std::forward<Value>(_value)) == std::end(_range), std::optional<size_t>())
+{
+	auto begin = std::begin(_range);
+	auto end = std::end(_range);
+	auto it = std::find(begin, end, std::forward<Value>(_value));
+	if (it == end)
+		return std::nullopt;
+	return std::distance(begin, it);
+}
+
+}
+
+/// @returns an std::optional<size_t> containing the offset of the first element in @a _range that is equal to @a _value,
+/// if any, or std::nullopt otherwise.
+/// Uses a linear search (``std::find``) unless @a _range is a container and provides a
+/// suitable ``.find`` function (e.g. it will use the logarithmic ``.find`` function in ``std::set`` instead).
+template<typename Range>
+auto findOffset(Range&& _range, std::remove_reference_t<decltype(*std::cbegin(_range))> const& _value)
+-> decltype(detail::findOffset(std::forward<Range>(_range), _value, 0))
```

Comment:
Why can't this be `auto`?

Reply:
This is what has SFINAE kick in for choosing the implementation... if this is too much, I can just remove the ``.find`` implementation though or even only implement the entire thing for ``std:::vector``, that's all I really need.

Reply:
ah ok!